### ISSUE TITLE
drive: multiple files of same name being created

### ISSUE THREAD

**Issue by benfry** [rclone/rclone#28]
Great tool! Thanks for all your work on this and making it available. 

Using `rclone sync` to a drive location, I'm seeing many copies of the same file/same name being created. Saw in the release notes that this was a fix in 1.06, but I've only ever used versions 1.07 and 1.09 (just downloaded 1.10 and will try that as well). 

Possible causes include: 
- interrupted transfers? (it's a huge directory and hasn't finished)
- maybe the deletions only occur after a sync is successful/complete?

Any thoughts?

**Comment by ncw** [comment]
rclone got a major fix to make drive transfers a lot more reliable in v1.09.  It is possible that the fix might be causing duplicated files though, and equally possible that the duplicated files were caused by an earlier version of rclone.

Can you see if you  can replicate it with v1.10 please?  As in you see v1.10 creating a duplicate file.

If you could send me the log rclone produces using the `-v` flag around the time you see a duplicated file that would be really helpful.

In the mean time I'll see if I can get rclone 1.10 to make duplicated files too.

Thanks

Nick

**Comment by ncw** [comment]
I've tested this with rclone v1.10 and I haven't seen it making duplicate files. I used my lots of small files test set which causes lots of retries which I thought might be the problem.

To answer your questions
  * interrupted transfers: if you restart the transfer then rclone should figure out what has and hasn't been uploaded and carry on
  * deletions do only occur after a sucessful sync, but rclone isn't expecting duplicate files so won't attempt to delete any of them

Let me know how your testing goes

Thanks

Nick

**Comment by benfry** [comment]
Honing in on a single folder and trying to sync it... Here's what the folder looks like on drive, after failed attempts over the last few weeks (note the duplicate names for several of the files). 

![rclone-dupes](https://cloud.githubusercontent.com/assets/1623101/6213045/b4f1228c-b5b9-11e4-86a7-37744730f1fb.png)

Re-running the sync for just that folder, it runs for a while on four files (each about 1-1.5 GB), but eventually times out and cancels all of them:

```
2015/02/16 12:38:17 Google drive root 'retired/clients/ge/aviation4/frames-every10th': Modify window is 1ms
2015/02/16 12:38:17 Google drive root 'retired/clients/ge/aviation4/frames-every10th': Building file list
2015/02/16 12:38:26 Google drive root 'retired/clients/ge/aviation4/frames-every10th': Waiting for checks to finish
2015/02/16 12:38:26 00.zip: Size and modification time the same (differ by 0, within tolerance 1ms)
2015/02/16 12:38:26 00.zip: Unchanged skipping
2015/02/16 12:38:26 02.zip: Size and modification time the same (differ by 0, within tolerance 1ms)
2015/02/16 12:38:26 04.zip: Size and modification time the same (differ by 0, within tolerance 1ms)
2015/02/16 12:38:26 04.zip: Unchanged skipping
2015/02/16 12:38:26 05.zip: Size and modification time the same (differ by 0, within tolerance 1ms)
2015/02/16 12:38:26 02.zip: Unchanged skipping
2015/02/16 12:38:26 06.zip: Size and modification time the same (differ by 0, within tolerance 1ms)
2015/02/16 12:38:26 06.zip: Unchanged skipping
2015/02/16 12:38:26 09.zip: Size and modification time the same (differ by 0, within tolerance 1ms)
2015/02/16 12:38:26 09.zip: Unchanged skipping
2015/02/16 12:38:26 05.zip: Unchanged skipping
2015/02/16 12:38:26 01.zip: Size and modification time the same (differ by 0, within tolerance 1ms)
2015/02/16 12:38:26 01.zip: Unchanged skipping
2015/02/16 12:38:26 12.zip: Size and modification time the same (differ by 0, within tolerance 1ms)
2015/02/16 12:38:26 12.zip: Unchanged skipping
2015/02/16 12:38:26 zip-em.sh: Size and modification time the same (differ by 0, within tolerance 1ms)
2015/02/16 12:38:26 zip-em.sh: Unchanged skipping
2015/02/16 12:38:26 Google drive root 'retired/clients/ge/aviation4/frames-every10th': Waiting for transfers to finish
2015/02/16 12:39:17 
Transferred:     58265600 Bytes ( 742.89 kByte/s)
Errors:                 0
Checks:                 9
Transferred:            0
Elapsed time:  1m16.592390836s
Transferring:  08.zip, 10.zip, 03.zip, 07.zip

2015/02/16 12:40:17 
Transferred:    142086144 Bytes (1015.83 kByte/s)
Errors:                 0
Checks:                 9
Transferred:            0
Elapsed time:  2m16.593337802s
Transferring:  03.zip, 07.zip, 08.zip, 10.zip

...lots more of the same...

2015/02/16 13:20:17 
Transferred:   5142712320 Bytes (1979.89 kByte/s)
Errors:                 0
Checks:                 9
Transferred:            0
Elapsed time:  42m16.592362404s
Transferring:  07.zip, 08.zip, 10.zip, 03.zip

2015/02/16 13:21:03 Google drive root 'retired/clients/ge/aviation4/frames-every10th': Error recived: googleapi: Error 401: Invalid Credentials, authError
2015/02/16 13:21:03 10.zip: Failed to copy: Upload failed: googleapi: Error 401: Invalid Credentials, authError
2015/02/16 13:21:03 10.zip: Removing failed copy
2015/02/16 13:21:17 
Transferred:   5272368843 Bytes (1982.91 kByte/s)
Errors:                 1
Checks:                 9
Transferred:            0
Elapsed time:  43m16.592358291s
Transferring:  03.zip, 07.zip, 08.zip, 10.zip

2015/02/16 13:21:23 Google drive root 'retired/clients/ge/aviation4/frames-every10th': Error recived: Post https://accounts.google.com/o/oauth2/token: dial tcp: lookup accounts.google.com: connection timed out
2015/02/16 13:21:23 10.zip: Failed to remove failed copy: Post https://accounts.google.com/o/oauth2/token: dial tcp: lookup accounts.google.com: connection timed out
2015/02/16 13:22:17 
Transferred:   5403507403 Bytes (1986.33 kByte/s)
Errors:                 1
Checks:                 9
Transferred:            1
Elapsed time:  44m16.592327945s
Transferring:  08.zip, 11.zip, 03.zip, 07.zip

2015/02/16 13:23:17 
Transferred:   5534579403 Bytes (1989.57 kByte/s)
Errors:                 1
Checks:                 9
Transferred:            1
Elapsed time:  45m16.593331043s
Transferring:  03.zip, 07.zip, 08.zip, 11.zip

2015/02/16 13:24:17 
Transferred:   5664668363 Bytes (1992.34 kByte/s)
Errors:                 1
Checks:                 9
Transferred:            1
Elapsed time:  46m16.592337997s
Transferring:  03.zip, 07.zip, 08.zip, 11.zip

2015/02/16 13:24:28 Google drive root 'retired/clients/ge/aviation4/frames-every10th': Error recived: googleapi: Error 401: Invalid Credentials, authError
2015/02/16 13:24:28 03.zip: Failed to copy: Upload failed: googleapi: Error 401: Invalid Credentials, authError
2015/02/16 13:24:28 03.zip: Removing failed copy
2015/02/16 13:24:28 Google drive root 'retired/clients/ge/aviation4/frames-every10th': Error recived: googleapi: got HTTP response code 404 with body: Not Found
2015/02/16 13:24:28 03.zip: Failed to remove failed copy: googleapi: got HTTP response code 404 with body: Not Found
2015/02/16 13:25:17 
Transferred:   5794267786 Bytes (1994.81 kByte/s)
Errors:                 2
Checks:                 9
Transferred:            2
Elapsed time:  47m16.592354661s
Transferring:  07.zip, 08.zip, 11.zip

2015/02/16 13:26:08 Google drive root 'retired/clients/ge/aviation4/frames-every10th': Error recived: googleapi: Error 401: Invalid Credentials, authError
2015/02/16 13:26:08 08.zip: Failed to copy: Upload failed: googleapi: Error 401: Invalid Credentials, authError
2015/02/16 13:26:08 08.zip: Removing failed copy
2015/02/16 13:26:08 Google drive root 'retired/clients/ge/aviation4/frames-every10th': Error recived: googleapi: got HTTP response code 404 with body: Not Found
2015/02/16 13:26:08 08.zip: Failed to remove failed copy: googleapi: got HTTP response code 404 with body: Not Found
2015/02/16 13:26:17 
Transferred:   5924284714 Bytes (1997.32 kByte/s)
Errors:                 3
Checks:                 9
Transferred:            3
Elapsed time:  48m16.592327211s
Transferring:  07.zip, 11.zip

2015/02/16 13:27:17 
Transferred:   6054701354 Bytes (1999.87 kByte/s)
Errors:                 3
Checks:                 9
Transferred:            3
Elapsed time:  49m16.592330029s
Transferring:  11.zip, 07.zip

2015/02/16 13:28:17 
Transferred:   6185511210 Bytes (2002.44 kByte/s)
Errors:                 3
Checks:                 9
Transferred:            3
Elapsed time:  50m16.592340633s
Transferring:  07.zip, 11.zip

2015/02/16 13:29:17 
Transferred:   6315927850 Bytes (2004.78 kByte/s)
Errors:                 3
Checks:                 9
Transferred:            3
Elapsed time:  51m16.592335557s
Transferring:  07.zip, 11.zip

2015/02/16 13:30:17 
Transferred:   6446573866 Bytes (2007.11 kByte/s)
Errors:                 3
Checks:                 9
Transferred:            3
Elapsed time:  52m16.592348462s
Transferring:  07.zip, 11.zip

2015/02/16 13:31:02 Google drive root 'retired/clients/ge/aviation4/frames-every10th': Error recived: googleapi: Error 401: Invalid Credentials, authError
2015/02/16 13:31:02 07.zip: Failed to copy: Upload failed: googleapi: Error 401: Invalid Credentials, authError
2015/02/16 13:31:02 07.zip: Removing failed copy
2015/02/16 13:31:02 Google drive root 'retired/clients/ge/aviation4/frames-every10th': Error recived: googleapi: got HTTP response code 404 with body: Not Found
2015/02/16 13:31:02 07.zip: Failed to remove failed copy: googleapi: got HTTP response code 404 with body: Not Found
2015/02/16 13:31:17 
Transferred:   6575510734 Bytes (2008.83 kByte/s)
Errors:                 4
Checks:                 9
Transferred:            4
Elapsed time:  53m16.592364375s
Transferring:  11.zip

2015/02/16 13:32:17 
Transferred:   6703764686 Bytes (2010.27 kByte/s)
Errors:                 4
Checks:                 9
Transferred:            4
Elapsed time:  54m16.592348358s
Transferring:  11.zip

2015/02/16 13:33:17 
Transferred:   6831789262 Bytes (2011.60 kByte/s)
Errors:                 4
Checks:                 9
Transferred:            4
Elapsed time:  55m16.592351568s
Transferring:  11.zip

2015/02/16 13:34:17 
Transferred:   6959748302 Bytes (2012.87 kByte/s)
Errors:                 4
Checks:                 9
Transferred:            4
Elapsed time:  56m16.592343089s
Transferring:  11.zip

2015/02/16 13:35:17 
Transferred:   7087838414 Bytes (2014.12 kByte/s)
Errors:                 4
Checks:                 9
Transferred:            4
Elapsed time:  57m16.592330152s
Transferring:  11.zip

2015/02/16 13:36:17 
Transferred:   7215830222 Bytes (2015.31 kByte/s)
Errors:                 4
Checks:                 9
Transferred:            4
Elapsed time:  58m16.592337792s
Transferring:  11.zip

2015/02/16 13:36:32 11.zip: Copied (new)
2015/02/16 13:36:32 Google drive root 'retired/clients/ge/aviation4/frames-every10th': Not deleting files as there were IO errors

Transferred:   7244781564 Bytes (2014.60 kByte/s)
Errors:                 4
Checks:                 9
Transferred:            5
Elapsed time:  58m31.861227027s

2015/02/16 13:36:32 *** Go routines at exit 9
```

**Comment by ncw** [comment]
Thanks that is really helpful.

I can see that the root cause of the duplicated files is the `googleapi: Error 401: Invalid Credentials, authError` error.

When that happens it fails to upload the file, and also fails delete the failed upload, leading to a duplicate.

I need to find out why those errors are happening which will require a bit of research - watch this space!

**Comment by benfry** [comment]
For what it's worth, they may be unrelated (the failure to copy and the duplicates). At least with 1.10, none of those four files that failed left behind a file with the same name, but several attempts to sync resulted in the same sequence of errors you see above. (Though I suspect that with earlier versions, they were failing a similar way and leaving behind the duplicates.) 

And in case it's useful, the timeout/bad credential errors come at what seems to be the *end* of the transfer. 

Thanks for looking into it!

**Comment by benfry** [comment]
Another data point, in case it's useful. Even attempting to sync a single file in that directory caused the same error:

```

2015/02/16 18:58:46 
Transferred:   1157858304 Bytes ( 376.29 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:  50m4.94123915s
Transferring:  03.zip

2015/02/16 18:59:20 Google drive root 'retired/clients/ge/aviation4/frames-every10th/03.zip': Error recived: googleapi: Error 401: Invalid Credentials, authError
2015/02/16 18:59:20 03.zip: Failed to copy: Upload failed: googleapi: Error 401: Invalid Credentials, authError
2015/02/16 18:59:20 03.zip: Removing failed copy
2015/02/16 18:59:20 Google drive root 'retired/clients/ge/aviation4/frames-every10th/03.zip': Error recived: googleapi: got HTTP response code 404 with body: Not Found
2015/02/16 18:59:20 03.zip: Failed to remove failed copy: googleapi: got HTTP response code 404 with body: Not Found
2015/02/16 18:59:20 Google drive root 'retired/clients/ge/aviation4/frames-every10th/03.zip': Not deleting files as there were IO errors

Transferred:   1228082111 Bytes ( 394.67 kByte/s)
Errors:                 1
Checks:                 0
Transferred:            1
Elapsed time:  50m38.700406368s
```

**Comment by ncw** [comment]
I think you may be right about these being two separate issues - the failure to auth and the duplicate files.  Could the duplicates have been made by an earlier version of rclone?

The auth errors look very much like these problems
  * http://stackoverflow.com/questions/23789284/resumable-upload-error-401
  * https://code.google.com/p/google-api-python-client/issues/detail?id=231

The google drive bug is fixed so it might be that I need to do a bit more work resuming the upload on auth failure.

**Comment by benfry** [comment]
Yeah, those look like the auth problems, certainly. (A little disappointed to see that it took 18 months to get a fix, and that Google's Drive client is using the v1 API.)

There are definitely duplicates from previous versions, but with another directory, I was also getting dupes with much smaller files (< 1Kb even) just last night with 1.10. Only thing that stuck out was that the files had either + or @ symbols in them (@synoresource files, it's backing up from a NAS), with could cause a string encoding issue (either in the code, or in the URLs), but that seems like a long shot. They may also have strange UNIX attribs, but I don't think rclone is paying attention to those.

One other thing on the dupes... They sometimes have different modification timestamps (though I don't think they've been modified), so if attributes are set after the file is uploaded and that fails, I wonder if things are getting stuck. Further, if the sync doesn't finish due to errors, the old files aren't cleaned out. But I haven't looked into your code to understand how it works through all those scenarios.

**Comment by ncw** [comment]
I have managed to reproduce the auth error problem and I'm working on a fix for that.  I'll probably break that into a separate issue.

I tried your suggestions with + or @ symbols for making duplicates but I haven't had any luck.  I'd hope that it isn't these funny characters causing a problem since the unit tests cover these tricky cases (I had to fix the google drive api to fix precisely this!) but you never know!

Do you think you could make a reproducer for me? Find a selection of files which you can sync to a fresh directory on drive and it will cause duplicates (every now and again is fine).  Perhaps it will need multiple syncs, or the timestamps changing on files or the length changing. You don't need to send the the files, all I need to know are their names and lengths (output of `rclone lsl /path/to/local/directory` would be perfect)

Thanks

Nick

**Comment by benfry** [comment]
Having trouble with developing a test case here. If I try to sync a smaller subset of files, things mostly go correctly: it skips files already transferred and reports no errors. However I do get "exit 9" or "exit 11" at the end of a sync.

In this case, the scenario is that paths passed on the command line are longer (since I'm looking at a subdirectory) and the paths reported for comparisons are shorter (because we're a couple directories deep)... Could there be an issue there? Or simply a directory depth issue?

As an aside, a successful (errors 0, checks 11240, but "routines at exit 9") sync to a folder that already has duplicates doesn't remove the duplicates, but from your early message it sounds like it's not designed to do that.

**Comment by ncw** [comment]
> However I do get "exit 9" or "exit 11" at the end of a sync.

If you are referring to the number of go routines at exit, then that is only for debugging, and you can ignore it.  It should never be huge, but the exact value is unimportant.  I put it in when I found a bug in the drive api [which I fixed](https://github.com/google/google-api-go-client/commit/967f22b3d5e3956a46b9ccfad98025dc3e3e707e)

> As an aside, a successful sync (errors 0, checks 11240, but "Go routines at exit 9") to a folder that already has duplicates doesn't remove the duplicates

It doesn't remove the duplicates, you are correct.  It assumes that duplicates aren't possible and the internal datastructures are arranged with that assumption!  It could be changed with a bit of work, but I'd rather not have the duplicates at all...

**Comment by madralphw** [comment]
Same problem uploading large files:

(50GB file)
2015/02/27 17:17:22 
Transferred:   51805750272 Bytes (12582.34 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:  1h7m0.837699975s
Transferring:  Blade2_C_Drive006.v2i

2015/02/27 17:17:54 Blade2_C_Drive006.v2i: Failed to copy: Upload failed: googleapi: Error 401: Invalid Credentials, authError

Transferred:   52428762515 Bytes (12631.21 kByte/s)
Errors:                 1
Checks:                 0
Transferred:            1
Elapsed time:  1h7m33.449969925s

**Comment by ncw** [comment]
@madralphw I'm working on that as we speak.  It affects all transfers which take longer than 1 hour.  Unfortunately neither of my first two attempts worked to fix the problem so cross fingers for attempt number 3!

**Comment by ncw** [comment]
I have factored the authError parts of this ticket into #33 to leave this one free for the duplicate file problem

**Comment by ncw** [comment]
I still haven't managed to replicate the duplicate file problem

However I've added checking in rclone for duplicate files so in sync/copy/check it will log a message about duplicate files so at least I can see when it is happening more easily.

**Comment by benfry** [comment]
Thanks for continuing to look into it. I'm trying 1.12 now and we'll see how it goes.

**Comment by benfry** [comment]
Still having some duplicates, though it seems to be far fewer than before. 

For one folder, it was a long path name and had a + in it, and by renaming the folder without the plus, I was able to re-transfer the whole folder (and it properly deleted the old folder, which had ~7 copies of each file). But based on what you've said, it sounds like neither the long name nor the plus is what's causing the trouble.

At least with the new version I can get predictable results: synching a folder that has multiple duplicates already may still create yet another duplicate, but subsequent attempts to sync that folder don't create yet another duplicate (and it faithfully reports “duplicate file detected” six or seven times for each).

**Comment by benfry** [comment]
Another data point... I have a folder with ~250 files that are several megabytes apiece. They're about 4–5 levels deep in the hierarchy of a directory that I'm trying to sync. At the moment, it's re-transferring the files, presumably creating the fourth or fifth set of duplicates. 

However, if I sync the two paths directly—as in, just the one folder with the duplicates—it recognizes all of the files as having already transferred (with "Duplicate file detected" messages and everything), and exits correctly. 

So in short, this creates duplicates:
`rclone sync /my/path/to/base/ gdf:base/`
and this does not:
`rclone sync /my/path/to/base/subfolder1/subfolder2/subfolder3/ gdf:base/subfolder1/subfolder2/subfolder3/`
Backing up one folder also works properly:
`rclone sync /my/path/to/base/subfolder1/subfolder2/ gdf:base/subfolder1/subfolder2/`

At the moment I'm suspicious that most things work as long as I have at least one subfolder after 'base' in the example above...

**Comment by ncw** [comment]
Thanks for that - I haven't done a lot of testing with deep folder paths, I'll try some more and see if I can make some duplicates.

**Comment by ncw** [comment]
@benfry I've fixed an issue which creates duplicate files on drive in #66 and released v1.14

I don't think that it is your issue but it might be worth a test!

**Comment by Kazu1226** [comment]
Hello from Japan. I still have the duplicate files problem. I use v.1.17 (64bit). 
Please resolve the problem!

**Comment by ncw** [comment]
@Kazu1226 thanks for the report.  If you have a log of duplicated files being created that would be really helpful.

**Comment by benfry** [comment]
@ncw thanks for the fix, will look into it.

**Comment by Kazu1226** [comment]
@ncw Here is some errors found in the log file.

Elapsed time:  59h40m8.3686986s
Transferring:  user/shinko/backup/SASANQUA/Backup Set 2013-04-15 083921/Backup Files 2013-04-16 120001/Backup files 3.zip, user/shinko/backup/SASANQUA/Backup Set 2013-04-15 083921/Backup Files 2013-04-16 120001/Backup files 4.zip, user/shinko/backup/SASANQUA/Backup Set 2013-04-17 120001/Backup Files 2013-04-17 120001/Backup files 1.zip, user/shinko/backup/SASANQUA/Backup Set 2013-04-15 083921/Backup Files 2013-04-16 120001/Backup files 5.zip



2015/07/13 08:33:30 user/shinko/backup/SASANQUA/Backup Set 2013-04-22 083812/Backup Files 2013-04-22 083812/Catalogs/Backup files 11.wbcat: Failed to copy: Upload failed: googleapi: Error 401: Invalid Credentials, authError
2015/07/13 08:33:30 user/shinko/backup/SASANQUA/Backup Set 2013-04-22 083812/Backup Files 2013-04-22 083812/Catalogs/Backup files 11.wbverify: Failed to copy: Upload failed: googleapi: Error 401: Invalid Credentials, authError
2015/07/13 08:33:30 user/shinko/backup/SASANQUA/Backup Set 2013-04-22 083812/Backup Files 2013-04-22 083812/Catalogs/Backup files 12.wbcat: Failed to copy: Upload failed: googleapi: Error 401: Invalid Credentials, authError
2015/07/13 08:33:30 user/shinko/backup/SASANQUA/Backup Set 2013-04-22 083812/Backup Files 2013-04-22 083812/Catalogs/Backup files 12.wbverify: Failed to copy: Upload failed: googleapi: Error 401: Invalid Credentials, authError
2015/07/13 08:34:09 
Transferred:   605230104932 Bytes (2735.45 kByte/s)
Errors:                78
Checks:             51943
Transferred:       110416
Elapsed time:  60h1m8.3691984s
Transferring:  user/shinko/backup/SASANQUA/Backup Set 2013-04-22 083812/Backup Files 2013-04-22 083812/Backup files 8.zip, user/shinko/backup/SASANQUA/Backup Set 2013-04-22 083812/Backup Files 2013-04-22 083812/Backup files 9.zip, user/shinko/backup/SASANQUA/Backup Set 2013-04-22 083812/Backup Files 2013-04-22 120001/Backup files 1.zip, user/shinko/backup/SASANQUA/Backup Set 2013-04-22 083812/Backup Files 2013-04-22 083812/Backup files 7.zip

**Comment by Kazu1226** [comment]
2015/07/13 01:09:43 user/satomi/xxxx/201404/20140428.doc: Received error: Upload failed - retry: googleapi: Error 403: User rate limit exceeded, userRateLimitExceeded - retrying 1/10
2015/07/13 01:09:44 user/satomi/xxxx/201404/20140428.doc: Received error: Upload failed - retry: googleapi: Error 403: User rate limit exceeded, userRateLimitExceeded - retrying 1/10

**Comment by Kazu1226** [comment]
2015/07/13 01:56:31 user/shinko/backup/SASANQUA/Backup Set 2012-12-26 120001/Backup Files 2012-12-27 120000/Backup files 7.zip: Received error: Chunk upload failed - retry: code=410, err=googleapi: Error 500: Internal Error, internalError - retrying 1/10
2015/07/13 01:56:40 user/shinko/backup/SASANQUA/Backup Set 2012-12-26 120001/Backup Files 2012-12-27 120000/Backup files 2.zip: Received error: Chunk upload failed - retry: code=410, err=googleapi: Error 500: Internal Error, internalError - retrying 1/10

**Comment by Kazu1226** [comment]
2015/07/13 04:25:40 user/shinko/backup/SASANQUA/Backup Set 2013-02-18 084229/Backup Files 2013-02-18 084229/Catalogs/Backup files 4.wbverify: Failed to copy: Upload failed: googleapi: Error 404: File not found: 0zdaDASDafdsa5djk3UzA, notFound
2015/07/13 04:25:41 user/shinko/backup/SASANQUA/Backup Set 2013-02-18 084229/Backup Files 2013-02-18 084229/Catalogs/Backup files 5.wbcat: Received error: Upload failed - retry: googleapi: Error 500: Internal Error, internalError - retrying 1/10
2015/07/13 04:25:51 user/shinko/backup/SASANQUA/Backup Set 2013-02-18 084229/Backup Files 2013-02-18 084229/Backup files 8.zip: Received error: Chunk upload failed - retry: code=410, err=googleapi: Error 500: Internal Error, internalError - retrying 1/10

**Comment by Kazu1226** [comment]
@ncw I am not sure if these errors are directly related to the duplicated files problem.

**Comment by jacobmcnamee** [comment]
It looks like the 401 Invalid Credentials errors typically occur very close to the oauth token expiry time. I believe this problem is caused by a race condition / time synchronization issue stemming from the goauth2 library.

If the local clock runs faster than the server clock, the token will first expire locally and be refreshed before it expires on the server. In this case, no requests are ever made with an expired token. In the reverse case, however, there is some window where the token is expired on the server but still seen as valid locally. Any requests made during this time will lead to 401 errors.

In addition to clock tolerance, other contributing factors could include network latency or abrupt time changes due to NTP.

There is a simple fix for this in the official oauth2 package, which you may want to migrate to eventually as goauth2 is deprecated. See https://github.com/golang/oauth2/commit/2b74ab268d043de5b0577a4773e67df08135e06d

**Comment by ncw** [comment]
@jacobmcnamee thanks - very interesting. I haven't had a good reason for switching to oauth2 but that sound like a good one.

**Comment by Kazu1226** [comment]
Thank you very much for working on the issue! I am looking forward to resolving this issue soon. :-)

**Comment by mbevin** [comment]
I have the duplicate files problem as well - in a folder I'm syncing almost every file is duplicated at least once, sometimes twice. I was wondering if the command-line options –checksum / –size-only might be worth trying to help avoid the issue? 

It appears I'm seeing the problem coming from rclone clients on both windows (version 1.22) + OSX (version 1.18).

And maybe rclone could also add a 'find+delete all such duplicates from folder' feature ;)

**Comment by mbevin** [comment]
Also note: I'm using rclone copy, with google drive

**Comment by ncw** [comment]
@mbevin I don't think `--checksum` or `--size-only` will help, but I'd be prepared to be proved wrong if you try it and find they make a difference! Thanks for the info -- Nick

**Comment by willpolley** [comment]
I'm having this problem as well. Not sure if it is tied to rate limits or connection issues.

Most of the dupes are small files. However, there are a few larger files which were duplicated.

**Comment by ncw** [comment]
What I could really do with is some kind volunteer to script rclone to do something like this
  * make a directory
  * upload some files
  * check for duplicates - exit if any found
  * remove the directory
  * start again

That can just be left running until it finds some duplicates so I can reproduce the problem.

**Comment by mbevin** [comment]
Tho I think you also want the script to repeat syncing/cloning multiple times before removing the directory, as it may be on latter syncs/clones that the duplicates are created (i.e. rclone somehow fails to recognize the file already exists and re-uploads it on the second run)?

**Comment by ncw** [comment]
@mbevin yes you are probably right

**Comment by ncw** [comment]
After a bit of investigation I think I've worked out what is happening. If anyone would like to have a go here is a workaround - using rclone 1.28 add `--low-level-retries 20` to your command line. This should stop the directory listings failing which is the root cause of the problem in my tests. Let me know how it goes!

**Comment by ncw** [comment]
Here is a beta with a fix for the directory listings to make it much more reliable.  This should eliminate the most common cause of duplicates.

http://pub.rclone.org/v1.28-6-gfdd4b4e%CE%B2/

**Comment by vishalmote** [comment]
Hello,

I have upgraded my rclone version to 1.28, and added above option as you have mentioned in command but after 6 to 7 successful sync days I got same failure error. Following is error. this is tail of error. There were thousands of lines of errors with mentioned as Duplicate file detected.

```
2016/04/19 04:47:08 source/web/app/storage/sessions/d7883be090891d1a38ecb05a0eb3dd089654a9a6: Duplicate file detected
2016/04/19 04:47:08 source/web/app/storage/sessions/927a578b406f282b72e5e2bac9098876f46435d2: Duplicate file detected
2016/04/19 04:47:08 source/web/app/storage/sessions/d02fa54ee35bc6be294229d2be07205a61b87ca2: Duplicate file detected
2016/04/19 04:47:08 source/web/app/storage/sessions/2047598524631273d6c2b939716e731e51c8120b: Duplicate file detected
2016/04/19 04:47:08 source/web/app/storage/sessions/.gitignore: Duplicate file detected
2016/04/19 04:47:08 Google drive root 'keurix_v4_backup': Waiting for checks to finish
2016/04/19 04:47:12 keurix/bamhc/drawings/032185d4d3904179c5ef5a04791f4219.pdf: Failed to copy: Post https://www.googleapis.com/upload/drive/v2/files?alt=json&uploadType=resumable: http: ContentLength=229 with Body length 0
2016/04/19 04:47:12 Google drive root 'keurix_v4_backup': Waiting for transfers to finish
2016/04/19 04:47:12 keurix/bamhc/drawings/de75f7c902f87811e5e78beec9a9ef03: Failed to copy: Post https://www.googleapis.com/upload/drive/v2/files?alt=json&uploadType=resumable: http: ContentLength=230 with Body length 0
2016/04/19 04:47:13 keurix/bamhc/drawings/1f6cefea4ec24727c6b72a6154f6c62a.pdf: Failed to copy: Post https://www.googleapis.com/upload/drive/v2/files?alt=json&uploadType=resumable: http: ContentLength=229 with Body length 0
2016/04/19 04:47:13 Attempt 3/3 failed with 3 errors

Transferred:    129403385 Bytes (   7.37 kByte/s)
Errors:                 3
Checks:            865706
Transferred:         1420
Elapsed time:  4h45m37.4s
```


Followings is what command I am using for it from shell script.

`/usr/sbin/rclone --low-level-retries 20 --contimeout=3h30m --timeout=0 --checkers=1 --transfers=1 --bwlimit 10M copy $DUMPDIR gdrive:keurix_v4_backup`

**Comment by ncw** [comment]
It would be interesting if you could estimate whether it is getting better or not.

The stats say rclone uploaded all but 3 files which is good.

I have another idea about the duplicate files - that the underlying cause may be duplicate directories.

I've put the drive fixes into v1.29 which I've just released and `dedupe` has a non-interactive mode too which should clear up the duplicates for you.

**Comment by vishalmote** [comment]
According to you, I dont have to worry about if few files not synced.  The data which I am syncing is very important, If single file wont get synced then my backup I am taking is useless for me as it is inconsistent data for me. Is there no solution that will do exact copy to drive without a error. I didn't get the above error. as it is saying there is 0 length body, but that file is in ok condition with having size in kbs.

**Comment by ncw** [comment]
@vishalmote I apologise, I didn't mean to suggest your data wasn't valuable, merely that 3 errors in 1420 is a better score than the number of duplicates suggest.

You can increase `--retries` to make it try harder to sync.  Also I note that you've set `--checkers` and `--transfers` very low - I suggest you set these higher again which will speed stuff up.

**Comment by bobbintb** [comment]
I started a sync job to Google Drive. It was going very slowly. It took a day to transfer an 11 GB file and then about 12 hours to transfer another 7 GB file. When I looked on Google Drive, there are two copies of each file. When I check the running sync job, it says 19GB have been transferred. I'm not sure where the issue lies but I'm willing to bet it is Google. There is no way I could have transferred that much in that amount of time. When I check the activity on Google, it looks like the duplicate file was uploaded a few minutes after the original, also impossible. Somehow, a duplicate is being created. It is still processing the two videos so maybe they are dummy files, I don't know. Anyone else come across this? No encryption enabled for this.
